src/webhook.rs

```diff
@@ -1,89 +1,72 @@
-extern crate crypto;
-extern crate url;
-extern crate base64;
-use self::crypto::hmac::Hmac;
-use self::crypto::sha1::Sha1;
-use self::crypto::mac::{MacResult,Mac};
-use hyper::server::request::Request;
-use hyper::header::Host;
-use hyper::uri::RequestUri::AbsolutePath;
-use hyper::method::Method::{Get,Post};
-use std::io::Read;
-use std::collections::HashMap;
+use crate::{Client, FromMap, TwilioError};
+use crypto::hmac::Hmac;
+use crypto::mac::{Mac, MacResult};
+use crypto::sha1::Sha1;
+use headers::{HeaderMapExt, Host};
+use hyper::{Body, Method, Request};
+use std::collections::BTreeMap;
 
-fn parse_object<T : ::FromMap>(args: &[(String,String)]) -> Result<Box<T>,::TwilioError> {
-    let mut m = HashMap::new();
-    for t in args {
-        m.insert(t.0.as_ref(),t.1.as_ref());
-    }
-    T::from_map(&m)
-}
-
-fn get_args(path: &str) -> Vec<(String,String)> {
+fn get_args(path: &str) -> BTreeMap<String, String> {
     let url_segments: Vec<&str> = path.split('?').collect();
     if url_segments.len() != 2 {
-        return vec![]
+        return BTreeMap::new();
     }
     let query_string = url_segments[1];
-    args_from_urlencoded(query_string)
+    args_from_urlencoded(query_string.as_bytes())
 }
 
-fn args_from_urlencoded(enc: &str) -> Vec<(String,String)> {
-    url::form_urlencoded::parse(enc.as_bytes()).into_owned().collect::<Vec<(String, String)>>()
+fn args_from_urlencoded(enc: &[u8]) -> BTreeMap<String, String> {
+    url::form_urlencoded::parse(enc).into_owned().collect()
 }
 
-impl ::Client {
-    pub fn parse_request<T : ::FromMap>(&self, req: &mut Request) -> Result<Box<T>,::TwilioError> {
-        let sig = match req.headers.get_raw("X-Twilio-Signature") {
-            None => return Err(::TwilioError::AuthError),
-            Some(d) => match d.len() {
-                1 => match base64::decode(&d[0]) {
-                    Ok(v) => v,
-                    Err(_) => return Err(::TwilioError::BadRequest),
-                },
-                _ => return Err(::TwilioError::BadRequest),
-            }
-        };
-        let mut bod = "".to_string();
-        req.read_to_string(&mut bod).unwrap();
-        let host: &str = match req.headers.get::<Host>() {
-            None => return Err(::TwilioError::BadRequest),
-            Some(h) => &h.hostname,
+impl Client {
+    pub async fn parse_request<T: FromMap>(
+        &self,
+        req: Request<Body>,
+    ) -> Result<Box<T>, TwilioError> {
+        let sig = req
+            .headers()
+            .get("X-Twilio-Signature")
+            .ok_or_else(|| TwilioError::AuthError)
+            .and_then(|d| match d.len() {
```

Comment:
D'oh! 🤦 That's a good catch. I wasn't aware that `get()` used to return an array of `HeaderValue`s instead of only one. You're correct that this would always return `TwilioError::BadRequest`, in that case. I think we can remove the `.len()` check and `match` statement, in that case. Thanks for pointing that out.

Reply:
Fixed in 2e8051b.